storage/transfermanager/downloader.go

```diff
@@ -155,46 +173,110 @@ func (d *Downloader) downloadWorker() {
 			break // no more work; exit
 		}
 
-		// TODO: break down the input into smaller pieces if necessary; maybe as follows:
-		// Only request partSize data to begin with. If no error and we haven't finished
-		// reading the object, enqueue the remaining pieces of work and mark in the
-		// out var the amount of shards to wait for.
-		out := input.downloadShard(d.client, d.config.perOperationTimeout)
-
-		// Keep track of any error that occurred.
-		if out.Err != nil {
-			d.error(fmt.Errorf("downloading %q from bucket %q: %w", input.Object, input.Bucket, out.Err))
-		}
-
-		// Either execute the callback, or append to results.
-		if d.config.asynchronous {
-			input.Callback(out)
+		out := input.downloadShard(d.client, d.config.perOperationTimeout, d.config.partSize)
+
+		if input.shard == 0 {
+			if out.Err != nil {
+				// Don't queue more shards if the first failed.
+				d.addResult(input, out)
+			} else {
+				numShards := numShards(out.Attrs, input.Range, d.config.partSize)
+
+				if numShards <= 1 {
+					// Download completed with a single shard.
+					d.addResult(input, out)
+				} else {
+					// Queue more shards.
+					outs := d.queueShards(input, out.Attrs.Generation, numShards)
+					// Start a goroutine that gathers shards sent to the output
+					// channel and adds the result once it has received all shards.
+					go d.gatherShards(input, outs, numShards)
```

Comment:
Do we need to block somewhere on this being completed if we're in async mode?

Reply:
We block in WaitAndClose until all results for each object that was added to the input are added to the result array or their callback is executed. 

I should note in the docs that in async mode WaitAndClose only finished until callbacks do (or we can start another goroutine to do the callback).